### DISCUSSION TITLE
Blazor sample?

### DISCUSSION THREAD

**Discussion by MisinformedDNA** [statiqdev/Discussions#133] [Q&A]
Do you have a sample app on how to use Statiq with Blazor?

Let's say I have a Blazor WASM site that loads content via an ID, I was wondering how I could use Statiq to generate those pages? Right now, I'm thinking I could just generate the JSON and then read the content from the JSON instead, but could I go further?

I would still like to use other Blazor features for dynamic content, so I'm not sure if I can get the best of both worlds somehow.

Thoughts?

**Comment by daveaglick** [comment]
I'd _love_ to figure out tighter Blazor support at some point. Blazor WASM is client-side so it makes sense that a static generator should, in theory, be able to support that model inline with other "normal" static content and pages. That said, my understanding is that there's a lot of magic that happens in the .NET compiler to light-up Blazor support and generate both the JavaScript bootstrapping code and the assemblies for a Blazor WASM site or component. We'd need to hook into that for a tight integration.

An alternative model could be similar to how I've seen Statiq used alongside ASPNET Core sites by using a single Solution with two different projects, one for Statiq and one for the ASPNET site. Then have Statiq generate it's content but set the output directory to somewhere the ASPNET site will serve from. I've even seen this model with Statiq being run in-process during app startup in the ASPNET site.

At this point though, I haven't used or explored Blazor very much at all since it's very early releases. It would probably take a fair amount of R&D to figure out how best to integrate more tightly, and I'm entirely focused on releasing Statiq Docs at the moment. Something like this is definitely on the backlog though, and if you blaze a trail and figure out some good approaches, even better!

**Reply by TheBrambleShark** [reply]
Currently I use Blazor for my website to handle some of the static scripting. That is, it's a script that runs at compile time to determine how something should look after it's written to html. For instance, I use this to figure out which navbar item should be selected based on the current path. This creates the illusion of the selection changing as you navigate the site but in reality the selected class is baked into the correct nav item.

You can see how this works for my [website](https://github.com/LuzFaltex/luzfaltex.com) by comparing the main and gh-pages branches.

**Reply by MisinformedDNA** [reply]
I'm still a Statiq beginner, but I don't even see where you are gathering the input files.

**Reply by TheBrambleShark** [reply]
Apologies, I only just saw this.
Statiq uses any file in the input directory as a source file.
If you use blazor files (`*.cshtml`) it will compile them to html first, executing any C# code as though you are rendering the page from an ASPX server.

This lets you calculate properties at compile time inside of code blocks, like so:
```cshtml
@{
    string name = "Foxtrek64";
}
<p>Hello, @name!</p>
```
```html
<p>Hello, Foxtrek64!</p>
```
The first sample is the raw cshtml and the second sample is the compiled code. Any active or interactive blocks would need to be javascript, at least currently.

I too would like to see the integration dave mentions. I think it'll be a huge help in some of the projects I have going.

**Reply by keywilliams** [reply]
@Foxtrek64 do you know if Statiq can read .razor files and export .html?
If yes, do you have some sample?
Thank You.

**Comment by keywilliams** [comment]
Hi @daveaglick, sorry for some mistakes.
I have the same @MisinformedDNA problem, then I would like to know if after this time, now it's possible to use Blazor (read .razor files and export to .html files) in Statiq?
If yes, do you have some sample?
Thank You.

**Reply by daveaglick** [reply]
> I would like to know if after this time, now it's possible to use Blazor (read .razor files and export to .html files) in Statiq

This feature isn't currently implemented, and I don't want to give you false hope - it likely won't be implemented for a while. There are a lot of complexities in Blazor compilation, not to mention the choice of different programming models (server-size rendering vs. fully client with WASM, etc.). If Statiq does support Blazor in the future, it'll almost certainly be using the fully client model, since that's essentially it's own little static generator. We'll need to integrate with MSBuild to make that work since most of the Blazor compilation and bootstrapping artifact generation happens as MSBuild tasks.

Even if/when Blazor is supported, I don't think it'll be strictly `.razor` -> `.html`. More like `.razor` -> a whole bunch of assemblies, some JavaScript, and other WASM bootstrapping code as generated by MSBuild. Those artifacts will then need to be integrated with the rest of the non-Blazor site (otherwise, there's no real benefit to using Statiq in the first place and MSBuild/VS/etc. could have just been used).

Long story short: I still think it's possible, and would like to support client-side Blazor as a component of a larger static site at some point, but it's not at the top of my priority list at the moment and given my availability recently, I don't see it being implemented outside a PR from the community any time soon.

**Reply by keywilliams** [reply]
@daveaglick thank you very much.
